### PR TITLE
Move experience stages to Lua

### PR DESCRIPTION
**Pull request by ranisalt**
Removed experience stages logic from C++, too.

The logic is a bit simpler in Lua. The complexity is O(quantity of stages + range of the largest stage), both are runtime constants. Moving to Lua code enables the possibility of more complex multiplier functions, such as the low level bonus introduced on recent Tibia updates as I have done here: ranisalt/forgottenserver@8552826 

If the lower constant complexity is really wanted, I can refactor it.

Closes #3049 #605

### CODE REVIEW

**Comment by djarek** [comment]
Nice. As for the complexity considerations - you shouldn't really worry. It's _incredibly unlinkely_ that your code has any significant impact on the overall performance. Last time I checked, around 60% of the total run time of the server is spent executing code related to spectator checking logic, so linear complexity is fine.

**Comment by mrhultin** [comment]
This is really neat. Should allow for both databasedriven stages aswell as 1.7x exp for example. Definatly going to use this.

**Comment by ranisalt** [comment]
@mrhultin it's perfectly possible as `player.lua` is executed at startup, just load from the database using `db` and you should be fine.

**Comment by mrhultin** [comment]
@ranisalt I have already done this, and it works like a charm I would say. I have not quite decided which way I like it more however, right now I use it compressed as Json in the row in the DB using server_config table as my table. This allows me to decompress the json quickly on the fly with PHP aswell.

**Comment by ranisalt** [comment]
Care to share the code? :D

**Comment by Raggaer** [comment]
How would external resources load the stages then? I think XML is better for this, how will AACs load your stages if they are on lua?

**Comment by ranisalt** [comment]
> How would external resources load the stages then? I think XML is better for this, how will AACs load your stages if they are on lua?

Using the status protocol, just like otserv lists do. There's even [a PHP library](https://github.com/renatorib/otinfo) to abstract that.

**Comment by ranisalt** [comment]
Is anyone interested? I'm not sure how people will prefer Lua over XML.

**Comment by WibbenZ** [comment]
Seems like a great thing to change, good job!
But you didnt think about the retards that can't read a change log to update servers from 0.x

I mean we should use both ways so ppl aren't afraid of using 1.x ofc

**Comment by ranisalt** [comment]
I'm not sure how can I maintain both behaviors. Maybe if the XML file does not exist, use Lua? But that wouldn't be really clean.

**Comment by WibbenZ** [comment]
It was more of a joke that I tought you would understand xD
It's abt actually trying new things and making noobs spend more than 10min to upgrade a server.

Tvh imo you should not keep the xml crap, if you are chaning smth and its for the better, why would you keep old crappy stuff for no good reason?

**Comment by ranisalt** [comment]
As a follower of http://semver.org/ that's actually something I'd hold for 2.0

**Comment by WibbenZ** [comment]
You never know, my guess is that 2.0(if we still have devs by then) will probbly be the same as now except some minor changes.

Sad to say thay since we could update alot of things if we stop all the "oh no the noobs cant do that"
But thats just my opinion tho :p

**Comment by gugahoa** [comment]
Use a flag on config.lua to choose which stages to use, Lua or XML, which defaults for XML
When we move to 2.0, remove the flag. (Add this to the roadmap so we don't forget, if merged?)

**Comment by marksamman** [comment]
You can already override Game.getExperienceStage if you rather implement it in Lua, or make changes to the Player.onGainExperience Lua event. Backwards compatibility is important and XML isn't bad for this kind of thing. The only thing that could use an improvement in the XML file is support for decimal precision.

**Comment by ranisalt** [comment]
Not only decimal precision, but you can really have a function instead of a table of values. I used the Lua system to have a very short and functional implementation of the low level bonus Tibia has: https://github.com/ranisalt/forgottenserver/commit/85528268409aedbd0cc7140996bf3f8773addeef

**Comment by Xaekai** [comment]
I don't like this with the configuration being a local to the tucked away player.lua.
```diff
- Game.getExperienceStage <-- this is unaddressed lost functionality -
```
If you're gonna move it to Lua, the configuration block should be a global setup, available to query by any script that may require it. If you're gonna make it difficult to use and begging for the server owner to rip it out and replace it with something else, why have it at all? Just my 2 cents. 🪙🪙

**Comment by ranisalt** [comment]
Now it has backwards compatibility with XML stages.

**Review by nekiro — DISMISSED** [review]

**Review by DSpeichert — DISMISSED** [review]

**Review by EPuncker — APPROVED** [review]

File: src/configmanager.cpp [hunk 1]
```diff
@@ -248,6 +328,19 @@ bool ConfigManager::getBoolean(boolean_config_t what) const
 	return boolean[what];
 }
 
+
+float ConfigManager::getExperienceStage(uint32_t level) const
+{
+	auto it = std::find_if(expStages.begin(), expStages.end(), [level](ExperienceStages::value_type stage) {
+		return level >= std::get<0>(stage) && level <= std::get<1>(stage);
+	});
```
Comment (nekiro) [inline src/configmanager.cpp:336]: actually, no newline after this? :(

**Review by nekiro — APPROVED** [review]

**Review by EPuncker — APPROVED** [review]

**Comment by Underewarrr** [comment]
Thats what i'm talking about, simple code for stages.
Goood JOB!

**Review by DSpeichert — APPROVED** [review]

**Pull request merged**